src/jit/gentree.h

```diff
@@ -2239,8 +2324,11 @@ class GenTreeUseEdgeIterator final
     AdvanceFn m_advance;
     GenTree*  m_node;
     GenTree** m_edge;
-    GenTree*  m_argList;
-    int       m_state;
+    union {
+        GenTree*         m_argList;
+        GenTreePhi::Use* m_phiUseList;
```

Comment:
From the first look I expected `m_phiUseList` to be `UseList`.

Reply:
More like `UseIterator`. I went back and forth on this but I think it needs to be a pointer due `GenTreeUseEdgeIterator`'s `==` operator
```C++
        return (m_node == other.m_node) && (m_edge == other.m_edge) && (m_argList == other.m_argList) &&
               (m_state == other.m_state);
```
This won't work correctly unless the 2 union members have the same layout/meaning. And even then it's very close to, if not right into, the undefined behavior land.

Reply:
I should add that in an ideal world `GenTreePhi::gtUses` should be `private`. But it seems that it requires quite a bit more work for not so many benefits, there are are only 3 remaining references to `gtUses` outside of `GenTreePhi`. At the end of the day I'm trying to make things reasonably better, not perfect.

Reply:
This seems a bit dicey to me - at least it might be good to make this a named union, and at least somewhat abstract the comparison that is done by the `==` operator, so that it's clear that these must be "compatible" in this way.

Reply:
Since you have the root node couldn't you check for the case where it's a PHI and handle it specially, say comparing the other union fields instead?

Reply:
Yep, I need to do something about this, it's a bit odd.

Reply:
> Since you have the root node couldn't you check for the case where it's a PHI and handle it specially, say comparing the other union fields instead?

That would work but I'm concerned that this will make `operator ==` even more complex than it is. Unfortunatelly, the C++ iterator model (and by extension range for) doesn't work so great when the state is complex. The repeated comparison with `end` can become costly.

I went for a more low tech approach - just use a `void*`. The few places that use this pointer can just cast to the right type.